### PR TITLE
refactor testing

### LINKED ISSUE #110: Testing should report accuracy across all files tested
We currently report a per-test-file entity accuracy. It would be nice to aggregate entity accuracy across multiple files, so you could get a single accuracy score for a range of tests that include flaky ones.

### PR DESCRIPTION
**Pull request by mgoldey**
- Refactor testing to use pytest
- Add nlprocessor integration for testing intent classification

Example invocation

```bash
python3 test_nlp.py -v --durations=50 --interpreter-directory=examples/directions/custom --tests examples/directions/tests.txt
```
```bash
python3 test_nlp.py -v --durations=50 --tests examples/transactions/tests.txt
```

Fixes #110. If don't change the interpreter, you can test and aggregate statistics for all appropriate test files.

### CODE REVIEW

**Review by yaroverg — APPROVED** [review]

File: launch.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+import docker
+import dotenv
+import requests
+from compose.cli.main import main as docker_compose
+from fire import Fire
+from mock import patch
+from requests.adapters import HTTPAdapter
+from urllib3.util.retry import Retry
+
+env = dotenv.dotenv_values("client.env")
```
Comment (sv-github) [inline launch.py:19]: :sunglasses:

File: util/regex_expander.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    >>> expand_these([["@num", "@letter"]])
+    [['@num', '@letter']]
+
+    This handles strings by themselves too now
+    >>> expand_these('a test')
+    [['a', 'test']]
+
+    Handle repeated patterns
+    >>> expand_these(["@num{1,2}", "@letter? @num{1,2}"])
+    [['@num'], ['@num', '@num'], ['@letter', '@num'], ['@letter', '@num', '@num']]
```
Comment (cbrochtrup) [inline util/regex_expander.py:89]: Thanks for the doctests. This makes the function's operation clear.

File: validate_regex.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    validate_entity_patterns(intent)
+    validate_compound_entities(intent)
+
+
+if __name__ == "__main__":
+    if len(sys.argv) < 2:
+        print("Please provide an input yaml file")
+        sys.exit(1)
+    for yaml_file in sys.argv[1:]:
+        validate_regex_in_yaml_file(yaml_file)
```
Comment (cbrochtrup) [inline validate_regex.py:100]: Any reason we didn't use Fire here?

**Review by cbrochtrup — COMMENTED** [review]
Awesome stuff! Pytest is more robust than a custom framework.

This was a lot to review, but I think I got it all.

File: cli_sdk.py [hunk 4]
```diff
@@ -1,84 +1,106 @@
...
-        validate_custom_directory(INTERPRETER_DIRECTORY),
-    )
+def start(interpreter_directory=os.environ.get("INTERPRETER_DIRECTORY",
+                                               "examples/calling_room")):
+    """
+    Start discovery
+    """
+    setup_discovery = partial(launch_docker_compose, "discovery")
+    validate_interpreter_directory(interpreter_directory)
+    setup_discovery(interpreter_directory)
```
[the remark this answers is not in the document]
Comment (mgoldey) [reply cli_sdk.py:27]: Partial to store arguments for later use has become something of a habit of mine. I'm happy to remove it here.

File: conftest.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+def check_discovery_setup(interpreter_directory, tests):
+    """
+    check for interpreter directory
+    """
+    if interpreter_directory:
+        validate_interpreter_directory(interpreter_directory)
+
+        # check all files
+        for yml_file in fnmatch.filter(interpreter_directory, ".yaml"):
```
[the remark this answers is not in the document]
Comment (mgoldey) [reply conftest.py:74]: Not enforced, but users are directed to .yaml per https://github.com/finos/greenkey-discovery-sdk/wiki/Definition-File-Overview

[the remark this answers is not in the document]
Comment (mgoldey) [reply conftest.py]: Fixed using glob instead

File: launch.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+        "nlprocessor.yaml",
+        "down",
+    ]
+    with patch.object(sys, "argv", args):
+        docker_compose()
+
+    for volume_name in [
+            "custom-discovery-interpreter",
+            "builtin-discovery-models",
+            "builtin-nlprocessor-models",
```
[the remark this answers is not in the document]
Comment (cbrochtrup) [reply launch.py:258]: Another option is to load these from the docker compose YAML files. I can see that causing errors, though, if someone fiddles with YAML.

**Pull request merged**